velox/vector/tests/utils/VectorMaker.h

```diff
@@ -588,21 +588,82 @@ class VectorMaker {
         arrays.push_back(std::nullopt);
         continue;
       }
-
       std::vector<std::optional<T>> elements;
-      for (const auto& element : arrayObject) {
-        if (element.isNull()) {
-          // Null element.
-          elements.push_back(std::nullopt);
-        } else {
-          elements.push_back(detail::jsonValue<T>(element));
-        }
-      }
+      elementVectorFromJsonHelper(elements, arrayObject);
+    }
+    return arrayVectorNullable<T>(arrays, arrayType);
+  }
 
-      arrays.push_back(elements);
+  template <typename T>
+  void elementVectorFromJsonHelper(
+      std::vector<std::optional<T>>& elements,
+      const folly::dynamic& arrayObject) {
+    // std::vector<std::optional<T>> elements;
+    for (const auto& element : arrayObject) {
+      if (element.isNull()) {
+        // Null element.
+        elements.push_back(std::nullopt);
+      } else {
+        elements.push_back(detail::jsonValue<T>(element));
+      }
     }
+  }
 
-    return arrayVectorNullable<T>(arrays, arrayType);
+  /// Creates an ArrayVector from a list of JSON arrays of arrays.
+  ///
+  /// JSON arrays of arrays can represent a null array, an empty array or array
+  /// with null elements.
+  ///
+  /// Examples:
+  /// [[1, 2], [2, 3, 4], [null, 7]]
+  /// [[1, 3, 7, 9], []]
+  /// [] - empty array of arrays
+  /// null - null array of arrays
+  /// [null]
+  /// [[]]
+  ///
+  /// @tparam T Type of array elements.
+  /// @param jsonArrays A list of JSON arrays. JSON array cannot be an empty
+  /// string.
+  template <typename T>
+  ArrayVectorPtr nestedArrayVectorFromJson(
+      const std::vector<std::string>& jsonArrays,
+      const TypePtr& arrayType = ARRAY(CppToType<T>::create())) {
+    std::vector<std::optional<std::vector<std::optional<T>>>> baseVector;
+    std::vector<vector_size_t> offsets;
+    std::vector<vector_size_t> nulls;
+    const std::vector<std::optional<T>> empty;
+    int offset = 0;
+    for (auto i = 0; i < jsonArrays.size(); i++) {
+      const auto& jsonArray = jsonArrays.at(i);
+      VELOX_CHECK(!jsonArray.empty());
+      const folly::dynamic arraysObject = folly::parseJson(jsonArray);
+      offsets.push_back(offset);
+      if (arraysObject.isNull()) {
+        // Null array.
+        nulls.push_back(i);
+        continue;
+      }
+      if (arraysObject.size() == 0) {
+        // Empty nested array
+        baseVector.push_back(empty);
+        offset++;
```

Comment:
Why do we increase the offset for empty array?

Reply:
Increasing the offset would be important to keep track of where the rows need to be created when making the nested array. To explain my thought process:
     
      "[[1, 2], [2, 3, 4], [null, 7]]",
      "[[1, 3, 7, 9], []]",
      "[]",

let's say this is the json I'm attempting to turn into an ArrayVector, the offsets keep track of where the rows in the nested array end. in this case, the offsets would be {0,3,5} where the rows would be made and the empty arrays would be counted here, so we need to add to offset when we encounter the empty arrays too.

Reply:
This is not right. Please, remove "if (arraysObject.size() == 0)" logic and update the test.

The last row in the example above should be an empty top-level array, not an array of size 1 with the only element being empty array.

See "review comments" commit in https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:arr_of_arr_changes?expand=1

Reply:
i see what you mean now. thank you for the clarification on this. The latest commit has all the top level array changes you suggested + other typo fixes